### PR TITLE
[Feature] Introduce executor factory provider

### PR DESCRIPTION
**Pull request by s7monk**
### Purpose

Close: #187.

Introduce `ExecutorFactoryProvider` to create the corresponding ExecutorFactory. If the task uses Spark engine, create a `SparkExecutorFactory`. If the task uses Flink engine, `FlinkSqlGatewayExecutorFactory` is created.

### Tests

- `ExecutorFactoryProviderTest#testGetExecutorFactory`

### API and Format

```
/** ExecutorFactoryProvider is responsible for providing the appropriate ExecutorFactory. */
public class ExecutorFactoryProvider {

    private final ExecutionConfig executionConfig;

    public ExecutorFactoryProvider(ExecutionConfig executionConfig) {
        this.executionConfig = executionConfig;
    }

    public ExecutorFactory getExecutorFactory(EngineType type) {
        switch (type) {
            case SPARK:
                throw new UnsupportedOperationException("Spark engine are not currently supported.");
            case FLINK_SQL_GATEWAY:
                return new FlinkSqlGatewayExecutorFactory(executionConfig.getSessionEntity());
            default:
                throw new IllegalArgumentException("Unsupported engine type: " + type);
        }
    }
}
```

### CODE REVIEW

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import lombok.Builder;
+import lombok.Getter;
+
+import java.util.Map;
+
+/** Configuration for execution context. */
+@Builder
+@Getter
+public class ExecuteConfig {
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:31]:
```suggestion
public class ExecutionConfig {
```

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+package org.apache.paimon.web.gateway.config;
+
+import org.apache.paimon.web.engine.flink.sql.gateway.model.SessionEntity;
+
+import lombok.Builder;
+import lombok.Getter;
+
+import java.util.Map;
+
+/** Configuration for execution context. */
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:28]: Is `ExecutionConfig` related with execution context?

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@Builder
+@Getter
+public class ExecuteConfig {
+
+    private boolean isStreaming;
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:33]: Could the execution config be used both streaming and batch?

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    private SessionEntity sessionEntity;
+
+    private Map<String, String> config;
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:37]:
```suggestion
    private Map<String, String> configs;
```

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/TaskType.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.gateway.enums;
+
+/** The {@code TaskType} enum defines the types of tasks that can be executed. */
+public enum TaskType {
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/TaskType.java:22]: IMO, the enum of task type is designed for the engine, not the implementation.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    private final ExecuteConfig executeConfig;
+
+    public ExecutorFactoryProvider(ExecuteConfig executeConfig) {
+        this.executeConfig = executeConfig;
+    }
+
+    public ExecutorFactory getExecutorFactory(TaskType type) {
+        switch (type) {
+                // Currently only supports Flink SQL Gateway.
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java:37]: Why does this introduce the provider of the `ExecutorFactory`? IMO, you wrap the executor too much and do not need to introduce provider for `ExecutorFactory`.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java (hunk 2 above)
Comment (s7monk) [reply paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:28]: yes.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java (hunk 1 above)
Comment (s7monk) [reply paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:31]: done.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java (hunk 3 above)
Comment (s7monk) [reply paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/config/ExecuteConfig.java:33]: Yes, we can support streaming and batch.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/TaskType.java (hunk 5 above)
Comment (SteNicholas) [reply paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/TaskType.java:22]: @s7monk, the enum could rename to `EngineType`.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            case SPARK:
+                throw new UnsupportedOperationException("Spark task are not currently supported.");
+            case FLINK:
+                return new FlinkSqlGatewayExecutorFactory(executeConfig.getSessionEntity());
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java:41]: Why does this only use `sessionEntity`?

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    private final ExecutionConfig executeConfig;
+
+    public ExecutorFactoryProvider(ExecutionConfig executeConfig) {
+        this.executeConfig = executeConfig;
+    }
+
+    public ExecutorFactory getExecutorFactory(TaskType type) {
+        switch (type) {
+                // Currently only supports Flink SQL Gateway.
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java:37]: Remove this comment.

**Comment by SteNicholas** [comment]
@s7monk, the enum has been changed, therefore please update the description of the pull request.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java (hunk 7 above)
Comment (s7monk) [reply paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java:41]: This class only uses sessionEntity here. It is currently only based on Flnk SQL Gateway. Other properties of ExecutionConfig are used in the server module.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/EngineType.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.web.gateway.enums;
+
+/** The {@code EngineType} enum defines the types of engines that can be supported. */
+public enum EngineType {
+    SPARK("SPARK"),
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/EngineType.java:23]:
```suggestion
    SPARK,
```

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/EngineType.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    FLINK("FLINK");
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/EngineType.java:24]: Ditto.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+    private final ExecutionConfig executionConfig;
+
+    public ExecutorFactoryProvider(ExecutionConfig executionConfig) {
+        this.executionConfig = executionConfig;
+    }
+
+    public ExecutorFactory getExecutorFactory(EngineType type) {
+        switch (type) {
+            case SPARK:
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProvider.java:37]: Remove this branch in this pull request, which could use `default` branch instead.

File: paimon-web-gateway/src/test/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProviderTest.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import static org.junit.jupiter.api.Assertions.assertSame;
+
+/** Test for {@link ExecutorFactoryProvider}. */
+public class ExecutorFactoryProviderTest {
+
+    @Test
+    public void testGetExecutorFactory() {
+        ExecutionConfig config =
+                ExecutionConfig.builder().sessionEntity(SessionEntity.builder().build()).build();
+        EngineType engineType = EngineType.fromValue("FLINK");
```
Comment (SteNicholas) [inline paimon-web-gateway/src/test/java/org/apache/paimon/web/gateway/provider/ExecutorFactoryProviderTest.java:37]: Add other engine type test case.

File: paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/EngineType.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package org.apache.paimon.web.gateway.enums;
+
+/** The {@code EngineType} enum defines the types of engines that can be supported. */
+public enum EngineType {
+    SPARK,
+    FLINK;
+
+    public static EngineType fromValue(String value) {
```
Comment (SteNicholas) [inline paimon-web-gateway/src/main/java/org/apache/paimon/web/gateway/enums/EngineType.java:26]:
```suggestion
    public static EngineType fromName(String name) {
```

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Pull request merged**